src/portable/raspberrypi/rp2040/rp2040_usb.c

```diff
@@ -126,8 +126,30 @@ void _hw_endpoint_start_next_buffer(struct hw_endpoint *ep)
 
     // PID
     val |= ep->next_pid ? USB_BUF_CTRL_DATA1_PID : USB_BUF_CTRL_DATA0_PID;
+
+#if TUSB_OPT_DEVICE_ENABLED
     ep->next_pid ^= 1u;
 
+#else
+    // For Host (also device but since we dictate the endpoint size, following scenario does not occur)
+    // Next PID depends on the number of packet in case wMaxPacketSize < 64 (e.g Interrupt Endpoint 8, or 12)
+    // Special case with control status stage where PID is always DATA1
+    if ( ep->transfer_size == 0 )
+    {
+      // ZLP also toggle data
+      ep->next_pid ^= 1u;
+    }else
+    {
+      uint32_t packet_count = 1 + ((ep->transfer_size - 1) / ep->wMaxPacketSize);
```

Comment:
haven't read the spec, but this just smells wrong... toggle the PID IF the transfer is an odd number number of packets total?

Reply:
Yup, the Data Toggle (PID) is toggled every packet. Following is the scheduling 108 bytes with endpoint with max packet size = 8. 
- Buffer 0 will start with 64 then 44 as last buf
- 64 is 8 packets starting with DATA1, each packet, DT toggled which mean after 8 packet, it is still DATA1. i.e at the 9th packet  (start of 44 bytes, transaction 87 in capture), the DT should remain the same (DATA1).

![image](https://user-images.githubusercontent.com/249515/120134401-375faf80-c1f8-11eb-8597-aae7e64409a2.png)

Reply:
PS: The log is there is on purpose, since I want to work more with rp2040 hcd/dcd, what I have in my mind is:
- Double buffering starting with host
- Make rp2040 common usb function more primitive, currently it handles also host and device specific code, making it more difficult to fix specific issue with either host/device (like above). Also it makes it difficult to follow in my taste.
- Host and device mode should be able to both enabled, and dynamically run depending on what the software want e.g using ID pin and/or VBUS sense to detect whether we attached to peer as host/device to initialize the correct mode.

Reply:
my question was whether the code was correct - as far as i can see the packet_count is only dependent on the total transfer size, and therefore does change), or is ep->transfer_size the size per packet?

Reply:
I don't quite understand your question could you give an example with above total transfer size = 108 bytes above to elaborate.  Here is how the transfer size is determined https://github.com/hathach/tinyusb/blob/host-rp2040/src/portable/raspberrypi/rp2040/rp2040_usb.c#L189

Reply:
I will try to explain as much as I could, as above example total transfer = 108 on endpoint with size = 8, which is broken into 2 sub transfers, both on buffer0 since we don't support double buffer just yet. 
- xfer1: transfer size = 64, start with DATA1, packet count = 8 
- xfer2: transfer size = 44, also start with DATA1, packet count = 5

if there is ever another xfer on the endpoint, it will start with DATA0 since the packet count is odd. Total transfer is broken in to 2 sub transfer size and therefore already taken into account when computing the next PID after this total transfer is complete. Hope that make sense, else I am happy to hear from your example.